[Crafting] Pluggable Widgets MCP

rahmanunver opened the pull request:
<!--
IMPORTANT: Please read and follow instructions below on how to
open and submit your pull request.

REQUIRED STEPS:
1. Specify correct pull request type.
2. Write meaningful description.
3. Run `pnpm lint` and `pnpm test` in packages you changed and make sure they have no errors.
4. Add new tests (if needed) to cover new functionality.
5. Read checklist below.

CHECKLIST:
- Do you have a JIRA story for your pull request?
    - If yes, please format the PR title to match the `[XX-000]: description` pattern.
    - If no, please write your PR title using conventional commit rules.
- Does your change require a new version of the widget/module?
    - If yes, run `pnpm -w changelog` or update the `CHANGELOG.md` and bump the version manually.
    - If no, ignore.
- Do you have related PRs in other Mendix repositories?
    - If yes, please link all related pull requests in the description.
    - If no, ignore.
- Does your change touch XML, or is it a new feature or behavior?
    - If yes, if necessary, please create a PR with updates in the documentation (https://github.com/mendix/docs).
    - If no, ignore.
 - Is your change a bug fix or a new feature?
      - If yes, please add a description (last section in the template) of what should be tested and what steps are needed to test it.
     - If no, ignore.
-->

<!--
What type of changes does your PR introduce?
Uncomment relevant sections below by removing `<!--` at the beginning of the line.
-->

<!-- No code changes (changes to documentation, CI, metadata, etc.)
<!---->

<!-- Dependency changes (any modification to dependencies in `package.json`)
<!---->

<!-- Refactoring (e.g. file rename, variable rename, etc.)
<!---->

<!-- Bug fix (non-breaking change which fixes an issue)
<!---->

<!-- New feature (non-breaking change which adds functionality)
<!---->

<!-- Breaking change (fix or feature that would cause existing functionality to change)
<!---->

<!-- Test related change (New E2E test, test automation, etc.)
<!---->

<!---
Describe your changes in detail.
Try to explain WHAT and WHY you change, fix or refactor.
-->

It's a WIP.
<!--
Please uncomment and fill in the following section
to describe which part of the package needs to be tested
and how it can be tested.
-->
<!--
### What should be covered while testing?
-->

rahmanunver committed:
feat(web-widgets): pluggable Widgets MCP for Agentic development

r0b1n commented on automation/mendix-widgets-copilot/src/property-engine.ts:
```diff
@@ -0,0 +1,508 @@
+        oldKey: string,
+        newKey: string
+    ): Promise<Omit<FileChange, "originalContent"> | null> {
+        const typingsPath = join(packagePath, "typings", `${this.getWidgetName(packagePath)}Props.d.ts`);
```
Typings are auto generated, so a safer way would be to run the generation using pwt when we add/remove properties.

r0b1n commented on automation/mendix-widgets-copilot/src/helpers.ts:
```diff
@@ -0,0 +1,221 @@
+        try {
+            const srcFiles = await readdir(srcPath);
+            for (const file of srcFiles) {
+                if (file.includes("editorConfig")) {
```
You can guess the name if you parse package json and get `widgetName` field + editorConfig + ".ts", same with preview files, widget xml, they all have the same pattern.

rahmanunver commented on automation/mendix-widgets-copilot/src/property-engine.ts:
```diff
@@ -0,0 +1,508 @@
+    }
+
+    // Rename operations (similar structure but with find/replace logic)
+    private async renamePropertyInXml(
```
I think we can modify and keep it as a guardrail, this can trigger a "Don't do it" logic we would like to provide. 
That guardrail can also be added into PRD, but doesn't an mcp level intervention sound safer? It is indeed a critical knowledge not to rename properties.

rahmanunver committed:
feat(mcp): implement MCP Sampling API and Prompts API with improved code structure

rahmanunver committed:
feat(mcp): add Sampling & Prompts APIs with improved architecture

Major enhancements to the MCP server for better AI assistance:

Sampling API (Context Awareness):
- Lets AI "preview" widget code before making changes
- Provides overview, properties, and runtime samples via Resources API
- Think: "Look under the hood before fixing"

Prompts API (Guided Workflows):
- 7 pre-built "recipes" for common tasks (add property, build, debug, etc.)
- Step-by-step guidance prevents AI mistakes
- Think: "IKEA instructions for widget development"

Architecture Improvements:
- Modular structure: sampling.ts, prompts.ts separate from helpers.ts
- Better separation of concerns
- Cleaner, more maintainable codebase

Documentation:
- Updated README with clear explanations
- Added usage examples for new features
- Included project structure overview

These changes make the AI smarter and less likely to break widgets by
providing context awareness and structured workflows.

rahmanunver committed:
chore(datagrid-web): rename to widgerts-mcp, .gitignore

rahmanunver committed:
feat(mendix-widgets-mcp): dockerize the server

rahmanunver committed:
build(mendix-pluggable-mcp): rebase and align deps
